arraykit.c

```diff
@@ -251,6 +261,716 @@ resolve_dtype_iter(PyObject *Py_UNUSED(m), PyObject *arg)
+    AK_GOTO_ON_NOT(cumsum, failure)
+
+    imask = PyNumber_Subtract(cumsum, PyLong_FromLong(1));
+    AK_GOTO_ON_NOT(imask, failure)
+
+    inv_idx = PyArray_Empty(
+            PyArray_NDIM(mask),             // nd
+            PyArray_DIMS(mask),             // dims
+            PyArray_DescrFromType(NPY_INT), // dtype
+            0);                             // is_f_order
+
+    if (PyObject_SetItem(inv_idx, ordered_idx, imask)) {
+        goto failure;
+    }
+
+    // 6. Pack it up in a tuple and return
+    PyObject* ret = PyTuple_Pack(2, filtered_arr, inv_idx);
+    AK_GOTO_ON_NOT(ret, failure)
+
+    PyMem_Free(mask_arr);
+
+    Py_DECREF(ordered_idx);
+    Py_DECREF(sar);
+    Py_DECREF(mask);
+    Py_DECREF(filtered_arr);
+    Py_DECREF(cumsum);
+    Py_DECREF(imask);
+    Py_DECREF(inv_idx);
+    return ret;
+
+failure:
+    Py_XDECREF(ordered_idx);
+    Py_XDECREF(sar);
+    Py_XDECREF(mask);
+    Py_XDECREF(filtered_arr);
+    Py_XDECREF(cumsum);
+    Py_XDECREF(imask);
+    Py_XDECREF(inv_idx);
+    return NULL;
+}
+
+static PyObject *
+AK_isin_array_dtype(PyArrayObject *array, PyArrayObject *other, int assume_unique)
+{
+    /* Algorithm:
+
+        array = np.ravel(array)
+
+        if not assume_unique:
+            array, rev_idx = np.unique(array, return_inverse=True)
+            other = np.unique(other)
+
+        concatenated = np.concatenate((array, other))
+
+        ordered_idx = concatenated.argsort(kind='mergesort')
+        sorted_arr = concatenated[ordered_idx]
+
+        flag = np.concatenate(((sorted_arr[1:] == sorted_arr[:-1]), [False]))
+
+        ret = np.empty(concatenated.shape, dtype=bool)
+        ret[ordered_idx] = flag
+
+        if assume_unique:
+            return ret[:len(array)]
+        else:
+            return ret[rev_idx]
+    */
+    // 0. Deallocate on failure
+    PyArrayObject* flattened_array = NULL;
+    PyObject *reverse_idx = NULL;
+    PyArrayObject* concatenated = NULL;
+    PyArrayObject *ordered_idx = NULL;
+    PyArrayObject* sorted_arr = NULL;
+    PyArrayObject* comparison = NULL;
+    PyArrayObject* ret = NULL;
+
+    // 1. Capture original array shape for return value
+    int array_ndim = PyArray_NDIM(array);
+    npy_intp* array_dims = PyArray_DIMS(array);
+    size_t array_size = PyArray_SIZE(array);
+
+    // 2. Ravel the array as we want to operate on 1D arrays only. (other is guaranteed to be 1D)
+    flattened_array = (PyArrayObject*)PyArray_Flatten(array, NPY_CORDER);
+    AK_GOTO_ON_NOT(flattened_array, failure)
+    Py_INCREF(flattened_array);
+
+    if (!assume_unique) {
+        PyObject* arr_and_rev_idx = AK_get_unique_arr_w_inverse(flattened_array);
+        PyArrayObject *raveled_array_unique = (PyArrayObject*)PyTuple_GET_ITEM(arr_and_rev_idx, 0);
+        AK_GOTO_ON_NOT(raveled_array_unique, failure)
+        Py_INCREF(raveled_array_unique);
+
+        reverse_idx = PyTuple_GET_ITEM(arr_and_rev_idx, 1);
+        AK_GOTO_ON_NOT(reverse_idx, failure)
```

Comment:
You leak `arr_and_rev_idx` and `raveled_array_unique` here.

Reply:
I need a bit more help explaining how this happening, and how to fix it?

Reply:
When you `goto failure` (inside of the macro), you still own references to `arr_and_rev_idx` and `arr_and_rev_idx`.

The solution is probably to replace the macro with an explicit `if (!reverse_idx) { ... }` where you `DECREF` those objects before the `goto`.